web3_proxy/src/frontend/authorization.rs

```diff
@@ -1083,6 +1093,34 @@ impl Web3ProxyApp {
         }
     }
 
+    /// Get the balance for the user.
+    ///
+    /// If a subuser calls this function, the subuser needs to have first attained the user_id that the rpc key belongs to.
+    /// This function should be called anywhere where balance is required (i.e. only rpc calls, I believe ... non-rpc calls don't really require balance)
+    pub(crate) async fn balance_checks(&self, user_id: u64) -> Web3ProxyResult<Decimal> {
+        // Downgrade the user if the balance is too low ...
+        if user_id == 0 {
+            Ok(Decimal::from(0))
+        } else {
+            self.user_balance_cache
+                .get_or_insert_async(&user_id, async move {
+                    let db_replica = self
+                        .db_replica()
+                        .web3_context("Getting database connection")?;
+
+                    let balance = balance::Entity::find()
+                        .filter(balance::Column::UserId.eq(user_id))
+                        .one(db_replica.conn())
+                        .await?
+                        .expect("related balance")
```

Comment:
Do not expect. There are tons of users in the db without balance rows and having this kind of expect was killing all of their requests before. This must be `unwrap_or_default`.